Adopt Safer CPP in RenderScrollbar.cpp

jelee53 opened the pull request:
#### 98067b6776eae3ae0d877d9340cc4a69d970da61
<pre>
Adopt Safer CPP in RenderScrollbar.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=304625">https://bugs.webkit.org/show_bug.cgi?id=304625</a>
<a href="https://rdar.apple.com/167057805">rdar://167057805</a>

Reviewed by Chris Dumez and Anne van Kesteren.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.</a>

No new tests needed.

* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/rendering/RenderScrollbar.cpp:
(WebCore::RenderScrollbar::RenderScrollbar):
(WebCore::RenderScrollbar::owningRenderer const):
(WebCore::RenderScrollbar::getScrollbarPseudoStyle const):
(WebCore::RenderScrollbar::updateScrollbarParts):
(WebCore::RenderScrollbar::updateScrollbarPart):
(WebCore::RenderScrollbar::paintPart):
(WebCore::RenderScrollbar::buttonRect const):
(WebCore::RenderScrollbar::trackRect const):
(WebCore::RenderScrollbar::trackPieceRectWithMargins const):
(WebCore::RenderScrollbar::minimumThumbLength const):
(WebCore::RenderScrollbar::opacity const):

Canonical link: <a href="https://commits.webkit.org/305365@main">https://commits.webkit.org/305365@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1cff05dbe24b4a5b9e53ec638d10b21ab497d68d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/138167 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/10532 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/49577 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/146244 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/91142 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/bfae20e0-f539-4aad-b33f-dbc1c03f5070) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/140041 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/11236 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/10686 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/105679 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/77093 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ea4ef482-40f0-4cdb-b52e-298a282e74d8) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/141113 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/8404 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/123871 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/86532 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b08487f4-379b-4298-9bf1-bd0458792b51) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7999 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/5760 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/6524 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/117409 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/42069 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/148952 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/10214 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/42626 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/114089 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/10231 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/8626 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/114424 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7947 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/120156 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/64938 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21283 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/10261 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/38095 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/9991 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/73828 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/10201 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/10052 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/rendering/RenderScrollbar.cpp:
```diff
@@ -76,12 +76,11 @@ RenderScrollbar::~RenderScrollbar() = default;
+        return m_owningFrame->ownerRenderer();
     }
     ASSERT(m_ownerElement);
     if (m_ownerElement->renderer())
```
You want to create a local CheckedPtr here instead.

jelee53 replied:
Had to do an `unsafeGet` to fix this. Please let me know if you think there's a better way.

annevk replied:
Here you haven't yet created a variable for `m_ownerElement->renderer()` but if you do and use that below to return `renderer->enclosingBox()` I don't think you'll have to use `unsafeGet()`.

annevk commented on Source/WebCore/rendering/RenderScrollbar.cpp:
```diff
@@ -182,15 +181,15 @@ void RenderScrollbar::updateScrollbarParts()
     int oldThickness = isHorizontal ? height() : width();
     int newThickness = 0;
-    RenderScrollbarPart* part = m_parts.get(ScrollbarBGPart);
+    CheckedPtr part = m_parts.get(ScrollbarBGPart);
```
Since part isn't used below, maybe move this assignment inside the conditional.

annevk commented on Source/WebCore/rendering/RenderScrollbar.h:
```diff
@@ -85,7 +86,7 @@ class RenderScrollbar final : public Scrollbar {
     RefPtr<Element> m_ownerElement;
 
-    WeakPtr<LocalFrame> m_owningFrame;
+    const WeakPtr<LocalFrame> m_owningFrame;
```
This isn't related to Safer CPP so probably best proposed separately.

annevk commented on Source/WebCore/rendering/RenderScrollbar.cpp:
```diff
@@ -75,13 +75,13 @@ RenderScrollbar::~RenderScrollbar() = default;
-        return currentRenderer;
+    if (RefPtr frame = m_owningFrame.get()) {
+        CheckedPtr currentRenderer = frame->ownerRenderer();
+        return currentRenderer.unsafeGet();
```
Here you can just return `frame->ownerRenderer();` without creating a local variable.

jelee53 replied:
I tried that in my original version but the safer CPP bot complainted about it.

annevk replied:
Are you sure? I think it complained about frame which you should keep, not ownerRenderer.

annevk commented on Source/WebCore/rendering/RenderScrollbar.cpp:
```diff
@@ -153,14 +153,14 @@ std::unique_ptr<RenderStyle> RenderScrollbar::getScrollbarPseudoStyle(ScrollbarP
+    scrollbarState.scrollCornerIsVisible = checkedScrollableArea()->isScrollCornerVisible();
     
-    std::unique_ptr<RenderStyle> result = owningRenderer()->getUncachedPseudoStyle({ pseudoElementType, scrollbarState }, &owningRenderer()->style());
+    std::unique_ptr<RenderStyle> result = checkedOwningRenderer()->getUncachedPseudoStyle({ pseudoElementType, scrollbarState }, &owningRenderer()->checkedStyle().get());
```
Since we null check owningRenderer above it seems better to obtain a local variable to it before that. At least that's more idiomatic WebKit style.

annevk commented on Source/WebCore/rendering/RenderScrollbar.cpp:
```diff
@@ -153,14 +153,14 @@ std::unique_ptr<RenderStyle> RenderScrollbar::getScrollbarPseudoStyle(ScrollbarP
     // causes visual artifact like non-repainted dirty region.
-    if (result && m_owningFrame && m_owningFrame->view() && !m_owningFrame->view()->isTransparent() && !result->hasBackground())
+    if (result && m_owningFrame && m_owningFrame->view() && !m_owningFrame->protectedView()->isTransparent() && !result->hasBackground())
         result->setBackgroundColor(Color::white);
```
Here it would be slightly better I think to create a nested if statement for the view() bit so you can create a local variable for view and avoid using protectedView().

annevk commented on Source/WebCore/rendering/RenderScrollbar.h:
```diff
@@ -45,6 +45,7 @@ class RenderScrollbar final : public Scrollbar {
     virtual ~RenderScrollbar();
 
     RenderBox* owningRenderer() const;
+    CheckedPtr<RenderBox> checkedOwningRenderer() const { return owningRenderer(); }
```
This might not be needed after you address my feedback above.

annevk reviewed (commented):
@cdumez might have additional thoughts as to whether this needs performance testing. Not sure how sensitive this code is.

annevk commented on Source/WebCore/rendering/RenderScrollbar.cpp:
```diff
@@ -152,15 +152,18 @@ std::unique_ptr<RenderStyle> RenderScrollbar::getScrollbarPseudoStyle(ScrollbarP
-        result->setBackgroundColor(Color::white);
+    if (result && m_owningFrame) {
+        RefPtr view = m_owningFrame->view();
+        if (view && view->isTransparent() && !result->hasBackground())
```
```suggestion
        if (view && !view->isTransparent() && !result->hasBackground())
```
Otherwise this would change behavior. Doesn't seem well-tested unfortunately.

jelee53 replied:
Can't believe that I missed that. Thanks for catching it!

annevk commented on Source/WebCore/rendering/RenderScrollbar.cpp:
```diff
@@ -152,15 +152,18 @@ std::unique_ptr<RenderStyle> RenderScrollbar::getScrollbarPseudoStyle(ScrollbarP
+    scrollbarState.scrollCornerIsVisible = checkedScrollableArea()->isScrollCornerVisible();
     
-    std::unique_ptr<RenderStyle> result = owningRenderer()->getUncachedPseudoStyle({ pseudoElementType, scrollbarState }, &owningRenderer()->style());
+    std::unique_ptr<RenderStyle> result = renderer->getUncachedPseudoStyle({ pseudoElementType, scrollbarState }, &renderer->checkedStyle().get());
```
Nit:
```suggestion
    std::unique_ptr<RenderStyle> result = renderer->getUncachedPseudoStyle({ pseudoElementType, scrollbarState }, renderer->checkedStyle().ptr());
```

cdumez commented on Source/WebCore/rendering/RenderScrollbar.cpp:
```diff
@@ -152,15 +152,18 @@ std::unique_ptr<RenderStyle> RenderScrollbar::getScrollbarPseudoStyle(ScrollbarP
-        result->setBackgroundColor(Color::white);
+    if (result && m_owningFrame) {
+        RefPtr view = m_owningFrame->view();
+        if (view && !view->isTransparent() && !result->hasBackground())
```
nit: Maybe we can move the ` && !result->hasBackground()` check to the earlier if check to avoid ref'ing the view unnecessarily.

cdumez reviewed (approved):
LGTM with nit.

jelee53 committed:
Adopt Safer CPP in RenderScrollbar.cpp
https://bugs.webkit.org/show_bug.cgi?id=304625
rdar://167057805

Reviewed by Chris Dumez and Anne van Kesteren.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.

No new tests needed.

* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/rendering/RenderScrollbar.cpp:
(WebCore::RenderScrollbar::RenderScrollbar):
(WebCore::RenderScrollbar::owningRenderer const):
(WebCore::RenderScrollbar::getScrollbarPseudoStyle const):
(WebCore::RenderScrollbar::updateScrollbarParts):
(WebCore::RenderScrollbar::updateScrollbarPart):
(WebCore::RenderScrollbar::paintPart):
(WebCore::RenderScrollbar::buttonRect const):
(WebCore::RenderScrollbar::trackRect const):
(WebCore::RenderScrollbar::trackPieceRectWithMargins const):
(WebCore::RenderScrollbar::minimumThumbLength const):
(WebCore::RenderScrollbar::opacity const):

Canonical link: https://commits.webkit.org/305365@main
